### PR TITLE
Unify CreateFile pinvoke in SMA

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Contributes to #18553.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:53]: I believe the intent of the name is that code doesn't call the C function directly

**Review by SteveL-MSFT — APPROVED** [review]
Looks fine to me, but there's a good amount of changes that should get another opinion

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (xtqqczze) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:53]: @SteveL-MSFT The `private` accessibility modifier should be sufficient.

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (xtqqczze) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:37]: If we are not going to use original naming convention, i.e. `FILE_FLAG_WRITE_THROUGH`, use C# convention.

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (jborean93) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:53]: We could have the one extern signature that returns nint and just have the outer methods wrap that in the specific safe handle types.

Comment (iSazonov) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:53]: The code was copied from .Net as well tested. Different teams create pinvoke code in .Net so it can have different styles. We have several PRs at the same time so it's also hard for us to follow one style. It is better to make a style correction later.

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:37]: It is a copy from old code. Really we have no need all the values and I kept them "just in case". Later we could correct style to follow .Net doc names.

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (jborean93) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:30]: If we are trying to unify this, isn't it ideal to do it properly first then migrate the code.

File: src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:1301]: If we set ownsHandle to false this will block dispose in NamedPipeClientStream - who will close the handle?

https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/SafeHandle.cs,237

Comment (jborean93) [reply src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:1301]: Oh my apologies, the Dispose() call is in a catch block not finally. I thought it was a finally block so this is all good.

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (jborean93) [inline src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:85]: Why does this return `nint` and not `SafePipeHandle`? I assume the purpose of having a separate call is to wrap it in a specific handle.

**Comment by iSazonov** [comment]
Since we do several PRs in parallel and copy code from both PowerShell and .Net, we already have a number of differences in style. This can be unified later if needed. I suppose it doesn't make sense to waste time adding unused constants or adding new enums because it's unlikely that this code will be used elsewhere. We aim to eliminate pinvokes in favor of .Net APIs.

**Comment by jborean93** [comment]
> We aim to eliminate pinvokes in favor of .Net APIs.

In this case dotnet do not seem like they are going to expose this in their own public API as per your recent issue https://github.com/dotnet/runtime/issues/79335. My concern is we are moving code that can be improved in several ways and by not updating it to a better standard we run the risk of it staying like that forever. I don't want to stand in the way of this PR but my concern is we will move this across as is then in the future people might be reluctant to change it because that's how the code is. Granted that's not a big concern but why not just do things properly once rather than twice.

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:85]: I preferred to make minimal changes to this code so as not to break anything.
As you can see it needs a thorough revision so it could probably be replaced by .Net code and most importantly it compiles for Unix when it should not.
It takes more knowledge to get the job done. Perhaps you could take a look at this if you are interested. That would be great!

Comment (jborean93) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:85]: If that's the case I wouldn't have a `CreateFileWithSafePipeHandle` just a generic `CreateFileW` that returns the `nint`. In the future when we are ready to expose something that will return a `SafePipeHandle` then we can look at adding this new pipe specific call.

**Comment by iSazonov** [comment]
> In this case dotnet do not seem like they are going to expose this in their own public API as per your recent issue [dotnet/runtime#79335](https://github.com/dotnet/runtime/issues/79335).

For this reason we will not be able to remove all pinvokes, but so where possible it should be done. And we need to keep working with the .Net team on new APIs.

> My concern is we are moving code that can be improved in several ways and by not updating it to a better standard we run the risk of it staying like that forever. I don't want to stand in the way of this PR but my concern is we will move this across as is then in the future people might be reluctant to change it because that's how the code is. Granted that's not a big concern but why not just do things properly once rather than twice.

Migrating to LibraryImport helps us at least make the main code more readable and remove auxiliary code from it. I'm very happy that this happens. For example, the FileSystemProvider file will become almost 10% smaller!
As for the style of pinvokes themselves, it doesn't bother me. For example, there will always be a debate which is better - to use constant names from documentation, or enums, or to combine these together. I have only one strict preference - to follow as close as possible to the .Net implementation, but as we know they don't have the same style there either. Moreover, they are still working on LibraryImport in .Net 8, so there will most likely be improvements there for us to follow.
It is more important to make the code better. This, too, can be a matter of discussion. For example, there is no point in implementing SafeHandle and adding a lot of allocations on hot path if we can easily open and close the handle in one place.
It's worth discussing anyway. To be specific, please open an issue(s) with a proposal for discussion. It can be about a specific code or a general style with examples. I am willing to participate in the discussions and actively promote your PRs.

File: src/System.Management.Automation/engine/Interop/Windows/FindFirstFile.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/Interop/Windows/FindFirstFile.cs:54]: I would rather use explicit names everywhere as I have caught a problem with them several times already :-) Moreover LibraryImport source generator always adds this explicitly.

**Comment by iSazonov** [comment]
> Looks fine to me, but there's a good amount of changes that should get another opinion

@SteveL-MSFT Do you ask about additional review? Who do you want to involve?

**Comment by SteveL-MSFT** [comment]
if @jborean93 approves, I think that would be sufficient

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (xtqqczze) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:30]: > Personally I recommend having FileAttributes as one parameter and another custom enum just for the flags and in the CreateFile call bitor them together

Since this is low-level internal code for P/Invoke, there's no particular reason to complicate matters by using the public FileAttributes Enum.

My preference is to stay as close a possible to the underlying API, see: https://github.com/dotnet/pinvoke/blob/main/src/Kernel32/Kernel32+CreateFileFlags.cs

File: src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs
Comment (xtqqczze) [reply src/System.Management.Automation/engine/Interop/Windows/CreateFile.cs:37]: Really, there is no need to follow the names in public .NET API.

**Comment by jborean93** [comment]
The only thing I think I really want to see is the removal of the pipe specific extern signatures. Currently it is adding no benefit over a straight call to `CreateFileW` as it isn't wrapping `SafePipeHandle` and the extended path/period end checks should not apply to pipes. I think we should just have 2 methods in this file

* `CreateFileW` that is the extern signature and returns `nint`
* `CreateFileWithSafeFileHandle` that does the max path and trailing period check that calls `CreateFileW` and wraps the return with `SafeFileHandle`

The current code that calls `CreateFileWithSafePipeHandle` can just call `CreateFileW` until we are ready to make a `CreateFileWithSafePipeHandle` (or the equivalent for pipes).

The other stuff I've mentioned in the review is mostly cosmetic and I think are fine to evolve over time in separate PRs.

**Comment by iSazonov** [comment]
>The only thing I think I really would like to see is the removal of the pipe specific extern signatures. Currently it is adding no benefit over a straight call to `CreateFileW` as it isn't wrapping `SafePipeHandle` and the extended path/period end checks should not apply to pipes. I think we should just have 2 methods in this file

I removed pipe path normalization since max pipe length is 256 https://learn.microsoft.com/en-us/windows/win32/ipc/pipe-names
I rename pipe helper (remove "Safe" from name) but keep the method since it hides two extra parameters in main code. I hope we remove the pinvoke at all in future.
@jborean93 Since you have great experience in this area could you please look and conclude if it is possible to get rid of this pinvoke at all with .Net code?

**Comment by jborean93** [comment]
At least for the pipes it should be possible to just use [NamedPipeClientStream](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipes.namedpipeclientstream?view=net-7.0). I'm not aware of any limitations it has over the `CreateFileW` call but as you've mentioned in another PR migrating to this might take more invasive changes we wanted to hold off from the PInvoke update PRs.

**Pull request merged**
